components/calendar/src/chinese_based.rs

```diff
@@ -309,6 +351,106 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
         debug_assert!(result < MAX_ITERS_FOR_MONTHS_OF_YEAR, "The given year was not a leap year and an unexpected number of iterations occurred searching for a leap month.");
         result
     }
+
+    /// Create a new arithmetic date from a year, month ordinal, and day with bounds checking; returns the
+    /// result of creating this arithmetic date, as well as a ChineseBasedCache - either the one passed in
+    /// optionally as an argument, or a new ChineseBasedCache for the given year, month, and day args.
+    pub(crate) fn new_from_ordinals(
+        year: i32,
+        month: u8,
+        day: u8,
+        cache: &ChineseBasedCache,
+    ) -> Result<ArithmeticDate<C>, CalendarError> {
+        let max_month = Self::months_in_year_cached(cache);
+        if month > max_month {
+            return Err(CalendarError::Overflow {
+                field: "month",
+                max: max_month as usize,
+            });
+        }
+
+        let max_day = Self::days_in_month(month, cache.new_year);
+        if day > max_day {
+            return Err(CalendarError::Overflow {
+                field: "day",
+                max: max_day as usize,
+            });
+        }
+
+        // Unchecked can be used because month and day are already checked in this fn
+
+        Ok(ArithmeticDate::<C>::new_unchecked(year, month, day))
+    }
+
+    /// Call `months_in_year_cached` on a `ChineseBasedDateInner`
+    pub(crate) fn months_in_year_inner(&self) -> u8 {
+        Self::months_in_year_cached(&self.1)
+    }
+
+    /// Return the number of months in a given year, which is 13 in a leap year, and 12 in a common year.
+    /// Also takes a `ChineseBasedCache` argument.
+    fn months_in_year_cached(cache: &ChineseBasedCache) -> u8 {
+        if cache.leap_month.is_some() {
+            13
+        } else {
+            12
+        }
+    }
+
+    /// Calls `days_in_month` on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_month_inner(&self) -> u8 {
+        Self::days_in_month(self.0.month, self.1.new_year)
+    }
+
+    /// Returns the number of days in the given `month` after the given `new_year`.
+    fn days_in_month(month: u8, new_year: RataDie) -> u8 {
+        let approx = new_year + ((month - 1) as i64 * 29);
+        let prev_new_moon = Self::new_moon_before((approx + 15).as_moment());
+        let next_new_moon = Self::new_moon_on_or_after((approx + 15).as_moment());
+        let result = (next_new_moon - prev_new_moon) as u8;
+        debug_assert!(result == 29 || result == 30);
+        result
+    }
+
+    /// Calls day_in_year_cached on an instance of ChineseBasedDateInner
+    pub(crate) fn days_in_year_inner(&self) -> u16 {
+        let next_new_year = Self::new_year_on_or_before_fixed_date(self.1.new_year + 400, None).0;
+        Self::days_in_year(self.1.new_year, next_new_year)
+    }
```

Comment:
Issue: The next new year is cached now so this function can be fast. That was a principal motivation behind adding it to the cache.